### DISCUSSION TITLE
Nested - DoE as inner method

### DISCUSSION THREAD

**Discussion by mikstery059** [snl-dakota/dakota#66] [Q&A] [answered]
Hi All,

I'm very new to Dakota, and DoE in general.  I was trying to do a nested study with a DACE method in the inner loop.  I was building off a file from a recent discussion post involving monte carlo nested inside monte carlo.  

Assuming the inner method iterates, say, 10 times per outer method iteration, how do I feed the mean response per each set of 10 responses to the outer loop (which uses it as its objective function)?  It seemed the example I was reading off of was using a method that naturally outputs the mean response and standard deviation via the primary response mapping, but I'm assuming DACE doesn't do that?  

Again I'm very new to Dakota, so I don't have a full grasp of the file/response pipelining...

V/R,
Michael

**Comment by jadamstephens** [comment, marked as answer]
Michael,

Dakota's dace method does not report moments and will not work as the inner method of a nested study. Which dace submethod were you hoping to use, and which Dakota method were you planning to use for the outer method?

Adam

**Reply by mikstery059** [reply]
Hi Adam,

Thank you for getting back to me so quickly!

I was thinking of employing oa_lhs for the inner method, and some global optimizer on the outside, maybe coliny direct?

Are moments not computed for DoE because they're undesirable?  Or is it just a lack of implementation?  I'm still new to the theory so I may be lacking some basic knowledge.  

I'm willing to just use UQ for the inner method if that's more desirable, I just wanted to pick something that suited my lack of knowledge of the distribution of the input space, and DoE seemed to fit that.

V/R,

Michael

**Reply by jadamstephens** [reply]
It's a lack of implementation due to intended usage.

Roughly speaking:

DACE is short for "design and analysis of computer experiments." These methods have their roots in traditional (physical) design of experiments. Design of experiments is aimed more at discovering important factors and the degree of their influence.

Moments are an "uncertainty quantification" result. If you look at the definitions of the moments, they include a weighting by probability mass or density, and our UQ methods use various strategies to "push forward" probability distributions on the inputs to the responses so that the moments can be computed.

There is some overlap in the way Dakota's dace and UQ methods work (for example: `dace lhs` and `sampling sample_type lhs` are provided by different libraries but work essentially the same way). But the DACE/UQ distinction is reflected in the way Dakota is structured, and it just doesn't compute UQ results for the DACE methods.

Long story short, if you need to do something like minimizing a mean, you should use `samplng`, not `dace oa_lhs`.

Adam

**Reply by mikstery059** [reply]
I understand.  Does this also mean that you can't calculate a value of your nonlinear constraint function in a custom way (i.e. a Python script) for the top-level optimizer if doing a nested study?  Or would such a constraint function only be able to work with the mean, response levels, etc?

**Comment by briadam** [comment]
Hi Michael -- It's possible an [optional interface pointer](https://snl-dakota.github.io/docs/6.18.0/users/usingdakota/reference/model-nested-optional_interface_pointer.html) will do what you want. It's intended to augment nested model final results, e.g., statistics, with outer loop evaluations of a simulation interface such as your Python nonlinear constraint evaluator. One example: https://github.com/snl-dakota/dakota/blob/devel/test/dakota_ouu1_tbch.in Documentation for optional interfaces is scant, so others may need to help, but just wanted to pitch the idea. -- Brian
